### PR TITLE
Database-adapter: commit optimizations

### PR DESCRIPTION
**Pull request by snazy**
* Do only fetch the HEAD of the target branch once
* Never try to fetch NO_ANCESTOR (it can never exist)
* Allow diffent sizes for the "embedded key list" (in a commit) and
  key-list-entities (separate rows)
* Nit: add commit-hash to conflict message

### CODE REVIEW

File: versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/DatabaseAdapterConfig.java [hunk 1]
```diff
@@ -111,6 +112,38 @@ default int getDefaultMaxKeyListSize() {
...
+   * Database adapter implementations that actually do have a hard technical or highly recommended
+   * limit on a maximum db-object / db-row size limitation should override this method and return a
+   * "good" value.
+   *
+   * <p>As for {@link #getMaxKeyListSize()}, this value must not be "on the edge" - means: it must
+   * leave enough room for database-serialization overhead * and similar.
+   *
+   * <p>Defaults to {@value #DEFAULT_MAX_KEY_LIST_ENTITY_SIZE}.
+   */
+  default int getDefaultMaxKeyListEntitySize() {
```
Comment (dimas-b) [inline versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/DatabaseAdapterConfig.java:143]: I'm not sure how it is going to work :thinking: In current config usage the `getMaxKeyListEntitySize` setting  apparently always overrides `getDefaultMaxKeyListEntitySize`... Is it worth having both of these settings? If we want the "default" size to be database-specific, should the code make sure `getDefaultMaxKeyListEntitySize` takes precedence?

File: versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/DatabaseAdapterConfig.java [hunk 2]
```diff
@@ -111,6 +112,38 @@ default int getDefaultMaxKeyListSize() {
     return DEFAULT_MAX_KEY_LIST_SIZE;
```
Comment (dimas-b) [inline versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/DatabaseAdapterConfig.java:112]: Do we need two database-level "hard" limit settings (wrt `getDefaultMaxKeyListEntitySize`)?

File: versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/spi/AbstractDatabaseAdapter.java [hunk 3]
```diff
@@ -725,6 +734,11 @@ protected final CommitLogEntry fetchFromCommitLog(OP_CONTEXT ctx, Hash hash) {
    * returned as {@code null}.
    */
   private List<CommitLogEntry> fetchMultipleFromCommitLog(OP_CONTEXT ctx, List<Hash> hashes) {
+    if (hashes.contains(NO_ANCESTOR)) {
+      // Do not try to fetch NO_ANCESTOR - it won't exist.
+      hashes = new ArrayList<>(hashes);
+      hashes.remove(NO_ANCESTOR);
```
Comment (dimas-b) [inline versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/spi/AbstractDatabaseAdapter.java:740]: javadoc: `The returned list must have exactly as many elements as in the parameter {@code hashes}.`

Aren't we violating the javadoc contract of this method here?

Comment (snazy) [reply versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/spi/AbstractDatabaseAdapter.java:740]: Duh! You're right!

File: versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/DatabaseAdapterConfig.java (hunk 2 above)
Comment (snazy) [reply versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/DatabaseAdapterConfig.java:112]: But yea - having a configurable default for another configurable doesn't really make much sense. Removed the configurable defaults.

Comment (dimas-b) [reply versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/DatabaseAdapterConfig.java:112]: Yes, that makes sense, but I guess that is controlled by the "non-default" parameters, right? My reading of the "default" parameters was that they define a hard limit for database-level entity sizes, which would normally be the same for any entity, I guess :thinking:

File: versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/DatabaseAdapterConfig.java (hunk 1 above)
Comment (dimas-b) [reply versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/DatabaseAdapterConfig.java:143]: Should the method then be on the `DatabaseAdapter` interface? Or perhaps of the adapter factory?

Comment (dimas-b) [reply versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/DatabaseAdapterConfig.java:143]: right now this value comes from the same source (e.g. `QuarkusVersionStoreAdvancedConfig`) as the other "key list" size, which makes it a bit confusing from my POV :thinking:

File: site/docs/try/configuration.md [hunk 4]
```diff
@@ -76,19 +76,21 @@ When setting `nessie.version.store.type=DYNAMO` which enables DynamoDB as the ve
...
+| `nessie.version.store.advanced.repository-id`                     |                     | `String` | Sets Nessie repository ID (optional). This ID can be used to distinguish multiple Nessie repositories that reside in the same storage instance.                                                                                       |
+| `nessie.version.store.advanced.parent-per-commit`                 | `20`                | `int`    | Sets the number of parent-commit-hashes stored in Nessie store.                                                                                                                                                                       |
+| `nessie.version.store.advanced.key-list-distance`                 | `20`                | `int`    | Sets the number "reachable" or "known" keys for each `CommitLogEntry`.                                                                                                                                                                |
```
[the remark this answers is not in the document]
Comment (snazy) [reply site/docs/try/configuration.md:83]: fixed

File: versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/DatabaseAdapterConfig.java (hunk 1 above)
Comment (snazy) [reply versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/DatabaseAdapterConfig.java:143]: (See below - will do in a separate PR)

Comment (snazy) [reply versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/DatabaseAdapterConfig.java:143]: But yes, it  (the hard limit) is an implementation concern and should be handled there.

**Review by dimas-b — DISMISSED** [review]
LGTM :+1:

**Review by dimas-b — APPROVED** [review]

**Pull request merged**